[Kernel] Adding basic Triton JitCache for triton_attn

bringlein opened the pull request:
In this PR we improve the performance of the V1 `triton_attn` backend. We do not change the triton kernels, but we reduce the well known launch overhead of triton kernels by caching the JIT artifacts of the decode kernel. 

## Performance 

All of the below results are for `meta-llama/Llama-3.1-8B-Instruct` on an NVIDIA H100 GPU. All experiments are done with `--no-enable-prefix-caching` and 

current upstream:
```
$ python benchmarks/benchmark_serving.py \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --dataset-name sharegpt \
    --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json 
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  25.54     
Total input tokens:                      215196    
Total generated tokens:                  197122    
Request throughput (req/s):              39.15     
Output token throughput (tok/s):         7718.08   
Total Token throughput (tok/s):          16143.83  
---------------Time to First Token----------------
Mean TTFT (ms):                          3572.95   
Median TTFT (ms):                        3427.32   
P99 TTFT (ms):                           6817.08   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          92.99     
Median TPOT (ms):                        53.98     
P99 TPOT (ms):                           239.31    
---------------Inter-token Latency----------------
Mean ITL (ms):                           44.60     
Median ITL (ms):                         32.69     
P99 ITL (ms):                            243.61    
==================================================
```

with this PR
```
$ python benchmarks/benchmark_serving.py \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --dataset-name sharegpt \
    --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  20.08     
Total input tokens:                      215196    
Total generated tokens:                  197466    
Request throughput (req/s):              49.80     
Output token throughput (tok/s):         9833.53   
Total Token throughput (tok/s):          20549.99  
---------------Time to First Token----------------
Mean TTFT (ms):                          3613.66   
Median TTFT (ms):                        3360.28   
P99 TTFT (ms):                           6966.82   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          88.58     
Median TPOT (ms):                        44.06     
P99 TPOT (ms):                           282.93    
---------------Inter-token Latency----------------
Mean ITL (ms):                           35.33     
Median ITL (ms):                         19.82     
P99 ITL (ms):                            234.20    
==================================================
```

to compare, using V1 `FlashAttention` backend:
```
$ python benchmarks/benchmark_serving.py \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --dataset-name sharegpt \
    --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  21.17     
Total input tokens:                      215196    
Total generated tokens:                  198001    
Request throughput (req/s):              47.23     
Output token throughput (tok/s):         9351.00   
Total Token throughput (tok/s):          19514.07  
---------------Time to First Token----------------
Mean TTFT (ms):                          3332.44   
Median TTFT (ms):                        3174.85   
P99 TTFT (ms):                           6246.38   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          83.21     
Median TPOT (ms):                        48.40     
P99 TPOT (ms):                           216.60    
---------------Inter-token Latency----------------
Mean ITL (ms):                           38.41     
Median ITL (ms):                         27.33     
P99 ITL (ms):                            220.79    
==================================================
```

So, this PR improves the performance of the V1 `triton_attn` backend by 27% and outperforms the FlashAttention-3 baseline by 5% for the serving benchmark. 

We are in the process of evaluating the performance on AMD GPUs.

## Correctness

Using the jitcache still produces correct results.

Using FlashAttention on H100 we see:

```
$ VLLM_USE_V1=1 lm_eval --model vllm --model_args pretrained=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.792|±  |0.0182|
|     |       |strict-match    |     5|exact_match|↑  |0.770|±  |0.0188|
```

with this PR, we see
```
$ VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 lm_eval --model vllm --model_args pretrained=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.79|±  |0.0182|
|     |       |strict-match    |     5|exact_match|↑  | 0.77|±  |0.0188|
```

## Details / How did we achieve this performance improvement

The launch overhead of triton kernels is a well known problem (see e.g. [1](https://github.com/triton-lang/triton/pull/3503), [2](https://github.com/triton-lang/triton/issues/2637), [3](https://github.com/triton-lang/triton/issues/6064)). Parts of the launch overhead comes from the fact that the triton JIT checks very carefully if an existing binary is safe to use. 

In many scenarios, these checks can be relaxed and check only a subset of the parameters. 
This PR adds such a cache with relaxed checks is implemented by `jitcache`. It is implemented as a decorator that could be used in front of the `triton.jit` decorator: 

```
@jitcache(
    check_keys=["USE_ALIBI_SLOPES", "SLIDING_WINDOW", "filter_by_query_len"],
)
@triton.jit
def kernel_paged_attention_.... 
```

As short description, the `jitcache` follows the steps of the triton JIT compiler to produce a binary, but does this for each version (indicated by `check_keys`) only once. For all consecutive invocations, only non-constant arguments are updated (and copied to GPU) skipping most parts of the compiler. 
A detailed usage description can be found in [here](https://github.com/IBM/triton-dejavu?tab=readme-ov-file#jitcache).

This reduces the launch overhead of the `paged_attention_2d` kernel of the triton backend from ~186us down to ~24us. 

## Discussion

We have added this new jit cache to `vllm/triton_utils/` because we expect that the vllm community prefers to have this rather as internal tool as an external dependency. However, we also published it as part of our [triton-dejavu](https://github.com/IBM/triton-dejavu) framework. So vllm could also import it from there, if this is preferred. 

Ideally, something like the `jitcache` could be added as feature to triton itself, but we expect this to maybe be a lengthier process. However, in all cases we would need to update the jitcache for every new triton release. 

<!--- pyml disable-next-line no-emphasis-as-heading -->

bringlein committed:
copying jit_cache, adapting jit cache for 2d kernel

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
some cleanup

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
formatting, typos...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
ruff....

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
adding assume const to jit cache

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
experimenting with static launch grid again

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
recovering good performance

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
going back to static launch grid

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
formatting...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
make type checking of key arguments more helpful

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
applying jit cache for prefix prefill

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
fmt & ruff

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
ci

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
remove changed requirements by mistake/pre-hook?

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
fmt...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
removing jit cache from prefix prefill again

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
cleanup

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein commented:
CC @SageMoore @tdoublep 

(I also removed the jitcache from `prefix_prefill` in this PR, to not conflict with https://github.com/vllm-project/vllm/pull/13305)

Sara-KS commented:
Testing this PR on MI-300X 
**Correctness:**

VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 VLLM_ROCM_CUSTOM_PAGED_ATTN=0 lm_eval --model vllm --model_args pretrained=/models/llama-3.1-8b/instruct/ --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.790|±  |0.0182|
|     |       |strict-match    |     5|exact_match|↑  |0.772|±  |0.0188|
```

**Performance:**
VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 VLLM_ROCM_CUSTOM_PAGED_ATTN=0 vllm serve /models/llama-3.1-8b/instruct/ --disable-log-requests --no-enable-prefix-caching

VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 VLLM_ROCM_CUSTOM_PAGED_ATTN=0 python benchmarks/benchmark_serving.py --model /models/llama-3.1-8b/instruct/ --dataset-name sharegpt --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json --ignore-eos

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  27.98     
Total input tokens:                      215196    
Total generated tokens:                  198343    
Request throughput (req/s):              35.73     
Output token throughput (tok/s):         7087.69   
Total Token throughput (tok/s):          14777.61  
---------------Time to First Token----------------
Mean TTFT (ms):                          8536.68   
Median TTFT (ms):                        7890.95   
P99 TTFT (ms):                           19628.92  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          31.62     
Median TPOT (ms):                        29.14     
P99 TPOT (ms):                           76.65     
---------------Inter-token Latency----------------
Mean ITL (ms):                           26.71     
Median ITL (ms):                         19.01     
P99 ITL (ms):                            85.35     
==================================================
```

The triton launch overhead does not appear to be a bottleneck on MI-300X. We are currently investigating this and may follow-up in a separate PR.

SageMoore commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,387 @@
+                self.update_args_index[arg_n] = i
+                self.non_const_vals_lst.append("dummy_value")
+            else:
+                self.non_const_vals_lst.append(assume_const_vals_dict[arg_n])
```
Assert that assume_const_vals_dict contains arg_n?

SageMoore commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,387 @@
+        self.non_const_vals_lst = []
+        self.update_args_index = {}
+        for i, arg_n in enumerate(self.non_const_arg_names):
+            if arg_n in update_only_arg_names:
```
I think some comments would be good here. Something like
```
            # If the argument can change each time the kernel is called, store a dummy value 
            # that will be set each time __call__ is called
            if arg_n in update_only_arg_names:
                self.update_args_index[arg_n] = i
                self.non_const_vals_lst.append("dummy_value")
            # else the argument is assumed to be constant and we can just store its initial value 
            else:
                self.non_const_vals_lst.append(assume_const_vals_dict[arg_n])
```

bringlein replied:
thanks, I added something similar.

SageMoore replied:
Is the main value of the `non_const_vals_lst` that it lets us do fewer copies in the __call__ function? It doesn't effect the caching at all right?

bringlein replied:
that is correct, yes

SageMoore commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,387 @@
+
+        self.cache_index_func = calc_cache_index
+        if len(check_keys) == 0:
+            self.cache_index_func = lambda ignore: "_default_"
```
This seems like a scary default. Wouldn't this just result in lots of false hits to the cache? I.E all functions mapping to the first function added?

SageMoore replied:
I think the current name is fine but we should be very clear about our expectations of users of the jit cache. Like naively I would somewhat expect key generation to be "transparent" from the users, meaning we can just make a key given the values of all constexpr and assumed_const args. Though, maybe there is a performance concern with making a key with so many values?

SageMoore replied:
Yeah, if that's the case, then I think that `cache_keys` should be a non-optional argument.

SageMoore reviewed (commented):
The infra looks good. I just want to make sure I'm understanding the key generation logic correctly.

bringlein commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,387 @@
+        cache_launch_grid=False,
+        assume_const=None,
+    ):
+        # we depend on the triton version, right now, 3.0 -- 3.2 are supported
```
Ok, so you prefer that the cache just "disables itself" and does nothing if the triton version is not supported (yet)? I could do that. Also I mean that for Triton 3.3 it would just do nothing and call regular JIT...

bringlein replied:
(In general I think we can support triton 3.3 as well, but that's not a focus right now)

SageMoore replied:
Yeah, I think that's the best solution for now. Crashing the whole process just because this caching system isn't supported yet seems like too much.

SageMoore replied:
How hard is 3.3 support? Since we've updated to torch 2.7 it would be good to get Triton 3.3 support here. Hopefully it just works?

bringlein commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,387 @@
+        self.base_fn = fn
+        while not inspect.isfunction(self.base_fn):
+            self.base_fn = self.base_fn.fn
+        self.cache_lock = cache_lock
```
right now, it is not used (only `_dynamic` is used in this PR), see also below.

SageMoore replied:
Since only `_dynamic` is used, can we remove this static vs dynamic distinction and simplify the code?

bringlein commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,387 @@
+        self.assume_const = assume_const
+        self.kernel_cache: dict[str, PreparedKernel] = {}
+
+        def calc_cache_index(kwargs):
```
Yes, you are right :). The developer needs to select these arguments based on her/his knowledge of the application. Put very basic, the `jitcache` trades safety in all scenarios and high launch overhead against a low launch overhead but reduced/relaxed safety checks applicable only to applications-specific use. It is then the job of the developers to ensure that the relaxed safety checks still hold for the particular application. 

In the case of the `paged_attention_2d` kernel, we assume that only the arguments `"USE_ALIBI_SLOPES", "SLIDING_WINDOW", "filter_by_query_len"` change during the lifetime of one vLLM instance (one model within vLLM, to be precise) (or at most these arguments, I actually think they barley change and we could still reduce the list). Other arguments like `num_heads` don't change during a models live time (at least, to my knowledge), hence we don't need to check them every time and realize "oh, didn't change...I just used some micro-seconds to ensure this again". 

I can try to mention this more explicitly in the doc strings.

bringlein replied:
I added some more explanations in the docstr of the decorator.

ProExpertProg replied:
Should we be more explicit with the JIT cache here and properly scope it to the model?

SageMoore replied:
I do think that we need to scope these caches to the model. Users are allowed to run multiple models without terminating the process.

tlrmchlsmth replied:
Are these caches scoped to the model in the current state of this PR?

bringlein committed:
address review comments

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,387 @@
+    :param assume_const: A list of parameters that are NOT marked as
+                         tl.constexpr but should be treated as constants in
+                         this kernel launch.
+    :param assume_const: list[str]
```
thanks, addressed.

bringlein commented on vllm/attention/ops/chunked_prefill_paged_decode.py:
```diff
@@ -324,8 +355,9 @@ def chunked_prefill_paged_decode(
             v_scale=v_scale,
         )
     else:
+        assert num_seqs <= 4096
```
thanks, addressed.

SageMoore replied:
Nit: minor mispelling "assume"

tlrmchlsmth replied:
Where does the 4096 come from in particular? Can we check `max_num_seqs` from the `SchedulerConfig` instead?

bringlein commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,387 @@
+
+        return prepared_kernel
+
+    def _run_static(self, *args, **kwargs):
```
I added some more explanations in the docstr of the decorator.

bringlein committed:
fix type hints

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
add transparency as fallback mode

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
CI whacamole

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
CI whacamole...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

SageMoore commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,418 @@
+        # we depend on the triton version, right now, 3.0 -- 3.2 are supported
+        if not (int(triton_version.split(".")[0]) == 3
+                and int(triton_version.split(".")[1]) <= 2):
+            logger.warning("JITCache is incompatible to installed Triton " \
```
Nit `logger.warning_once`

SageMoore reviewed (commented):
Apologies for the delay @bringlein. I've left another round of comments. In general this PR looks good, but I really want to make sure that we don't get any accidental false positives :).

Additionally, I think it would be good to add an "escape hatch" that disables the jit cache all together. I don't have a super strong opinion on how we do this. I think an env var that eventually becomes a full config option once this matures a bit more would be fine.

bringlein committed:
Merge branch 'main' into ngl_jit_cache_pr

bringlein committed:
adding triton 3.3 support

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
Merge branch 'main' into ngl_jit_cache_pr

bringlein committed:
fixing triton 3.3 support (1/x); add support for unified kernel

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
fixing triton 3.3 support (2/2)

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
cleanup and add env var

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
adding assume_const

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
make argument passing (slightly) faster

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
Merge branch 'main' into ngl_jit_cache_pr (moving envs content)

bringlein committed:
fixing env var merge conflict

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
adding attention metadata specific for triton_backend

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
fixing env file again

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
Revert "adding attention metadata specific for triton_backend"

This reverts commit 450770c49d17cc5dd89ebdc4f3cedbc70eb374b5.

bringlein committed:
more elegant fix on dependency of flash attention

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
thrid way to un-break triton backend

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
CI...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein commented:
some updates on this PR: 
- The `jitcache` in this PR is now compatible with triton 3.3 (and only 3.3, for backward compatibility, please use [triton-dejavu](https://github.com/IBM/triton-dejavu/tree/main?tab=readme-ov-file#jitcache). 
- The jitcache is disabled by default, to enable it the environment variable `VLLM_TRITON_ENABLE_JITCACHE=1` needs to be set. 
- I updated the unified triton attention kernel from PR https://github.com/vllm-project/vllm/pull/16828 to use the updated jitcache. 
- I fixed the broken attention backend on cuda platform (which was broken due to the dependency on the `FlashAttentionMetadataBuilder`). 

From this PR, two todos remain:
1. With triton 3.3, we see a performance regression for the triton backend. Details see below. However, this regression is independent of the jitcache and is therefore addressed in another PR. 
2. The triton backen is still depending on the attention metadata from the V1 flash attention backend, which should be fixed and made fully independent (maybe including to move some of the flash attention metadata functionalities to a utility file). However, also this is independent of the jitcache and should be addressed in another PR. 

### Correctness: 

```
$ VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 VLLM_TRITON_ENABLE_JITCACHE=1 lm_eval --model vllm --model_args pretrained=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500
vllm (pretrained=meta-llama/Llama-3.1-8B-Instruct), gen_kwargs: (None), limit: 500.0, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.798|±  |0.0180|
|     |       |strict-match    |     5|exact_match|↑  |0.782|±  |0.0185|
```


### Performance: 

### tracing

If analyzing vllm profiles, the jitcache functions as expected and reduces the triton launch overhead from 148us to 26us:
before:
![image](https://github.com/user-attachments/assets/393c7724-e786-400b-a385-1221c12081cf)
after:
![image](https://github.com/user-attachments/assets/ac275602-77ae-4bad-83ad-235b938a6a0b)


### H100

With pytorch 2.7 and triton 3.3, the performance of the triton attention with this PR on an H100 drops to:
```
VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 VLLM_TRITON_ENABLE_JITCACHE=1 vllm serve meta-llama/Llama-3.1-8B-Instruct --disable-log-requests --no-enable-prefix-caching

VLLM_USE_V1=1 python benchmarks/benchmark_serving.py --model meta-llama/Llama-3.1-8B-Instruct --dataset-name sharegpt --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json

============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  30.49     
Total input tokens:                      215196    
Total generated tokens:                  197999    
Request throughput (req/s):              32.80     
Output token throughput (tok/s):         6493.77   
Total Token throughput (tok/s):          13551.56  
---------------Time to First Token----------------
Mean TTFT (ms):                          4738.41   
Median TTFT (ms):                        4668.27   
P99 TTFT (ms):                           9096.74   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          119.62    
Median TPOT (ms):                        65.38     
P99 TPOT (ms):                           322.74    
---------------Inter-token Latency----------------
Mean ITL (ms):                           53.69     
Median ITL (ms):                         36.63     
P99 ITL (ms):                            330.55    
==================================================
```

with `VLLM_TRITON_ENABLE_JITCACHE=0` the performance is:
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  31.05     
Total input tokens:                      215196    
Total generated tokens:                  198064    
Request throughput (req/s):              32.21     
Output token throughput (tok/s):         6378.75   
Total Token throughput (tok/s):          13309.25  
---------------Time to First Token----------------
Mean TTFT (ms):                          4927.07   
Median TTFT (ms):                        4718.52   
P99 TTFT (ms):                           9635.06   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          124.71    
Median TPOT (ms):                        67.28     
P99 TPOT (ms):                           365.14    
---------------Inter-token Latency----------------
Mean ITL (ms):                           55.10     
Median ITL (ms):                         37.23     
P99 ITL (ms):                            338.55    
==================================================
```
(so roughly only 2% increase in total performance due to the jitcache). 

#### MI300


With pytorch 2.7 and triton 3.3, the performance of the triton attention with this PR on an H100 drops to:
```
VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 VLLM_TRITON_ENABLE_JITCACHE=1 vllm serve mistralai/Mistral-Small-24B-Instruct-2501 --disable-log-requests --no-enable-prefix-caching

VLLM_USE_V1=1 python benchmarks/benchmark_serving.py --model mistralai/Mistral-Small-24B-Instruct-2501 --dataset-name random --request-rate 9 --random-input-len 1000 --random-output-len 100 --num-prompts 1080 --ignore-eos --seed 9

============ Serving Benchmark Result ============
Successful requests:                     1080      
Benchmark duration (s):                  154.60    
Total input tokens:                      1078920   
Total generated tokens:                  108000    
Request throughput (req/s):              6.99      
Output token throughput (tok/s):         698.57    
Total Token throughput (tok/s):          7677.32   
---------------Time to First Token----------------
Mean TTFT (ms):                          10223.16  
Median TTFT (ms):                        9194.06   
P99 TTFT (ms):                           22111.30  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          710.73    
Median TPOT (ms):                        801.62    
P99 TPOT (ms):                           1004.43   
---------------Inter-token Latency----------------
Mean ITL (ms):                           710.73    
Median ITL (ms):                         1016.49   
P99 ITL (ms):                            1070.81   
==================================================
```

with `VLLM_TRITON_ENABLE_JITCACHE=0` the performance is:
```
============ Serving Benchmark Result ============
Successful requests:                     1080      
Benchmark duration (s):                  154.65    
Total input tokens:                      1078920   
Total generated tokens:                  108000    
Request throughput (req/s):              6.98      
Output token throughput (tok/s):         698.34    
Total Token throughput (tok/s):          7674.80   
---------------Time to First Token----------------
Mean TTFT (ms):                          10475.19  
Median TTFT (ms):                        9509.31   
P99 TTFT (ms):                           22325.98  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          708.41    
Median TPOT (ms):                        798.45    
P99 TPOT (ms):                           1001.22   
---------------Inter-token Latency----------------
Mean ITL (ms):                           708.41    
Median ITL (ms):                         1013.82   
P99 ITL (ms):                            1067.09   
==================================================

```

cyang49 committed:
making jitcache safe to use with autotuner

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
CI whacamole...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

tlrmchlsmth commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,417 @@
+        self.cache_launch_grid = cache_launch_grid
+        self.concrete_grid = (None, None, None)
+        if cache_launch_grid:
+            grid_0 = grid_example[0]
+            grid_1 = grid_example[1] if self.grid_size > 1 else 1
+            grid_2 = grid_example[2] if self.grid_size > 2 else 1
+            self.concrete_grid = (grid_0, grid_1, grid_2)
```
what if `self.grid_size` is greater than 3?

bringlein replied:
That is prohibited by triton and will throw an error even before the jitcache is being called.

tlrmchlsmth commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,417 @@
+        if self.cache_launch_grid:
+            grid_0, grid_1, grid_2 = self.concrete_grid
+        else:
+            if self.grid_is_callable:
+                grid = kwargs["grid"](kwargs)
+            else:
+                grid = copy.deepcopy(kwargs["grid"])
```
Is there a downside to always using `grid = copy.deepcopy(kwargs["grid"])`? It would be better to simplify the code if this isn't a significant optimization

ProExpertProg replied:
The first option calls `grid` so it's not the same code

tlrmchlsmth commented on vllm/triton_utils/jit_cache.py:
```diff
@@ -0,0 +1,417 @@
+        # assert no config pre-hook
+        assert "pre_hook" not in kwargs or kwargs["pre_hook"] is None
+
+        # print(f"my lock: {self.cache_lock.is_locked}")
```
nit: remove cruft before landing

tlrmchlsmth committed:
fixup spelling in a few spots

Signed-off-by: Tyler Michael Smith <tysmith@redhat.com>

tdoublep committed:
Merge branch 'main' into ngl_jit_cache_pr

tdoublep committed:
Added support for specialization.

Co-authored-by: Burkhard Ringlein <ngl@zurich.ibm.com>
Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

tdoublep commented:
I worked a bit on this PR today to:
- Resolve conflicts with recent changes on main
- Add support for handling specialization of non-const integer parameters (e.g.  num_seqs).

I don't have write perms on @bringlein's fork so I pushed the latest code to my fork here:
https://github.com/tdoublep/vllm/tree/ngl_jit_cache_pr

I've done quite a lot of benchmarking on H100 and unfortunately I'm not really seeing any significant performance improvement anymore. I believe the "big" performance improvement we were seeing in the past (with Triton 3.2) was an artifact of not handling the specialization correctly. 

These are the results of running the ShareGPT serving benchmark on H100, repeated 10 times, with and without jitcache:

<img width="734" alt="image" src="https://github.com/user-attachments/assets/e6e7fc61-7b74-4af5-b595-a1531e28ad5b" />


The only benefit seems to be that the performance is more "stable" (standard deviation is lower).

bringlein committed:
Merge branch 'main' into ngl_jit_cache_pr

bringlein committed:
minor cleanup; remove copy of launch grid

bringlein committed:
improve docstring

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
Merge branch 'main' into ngl_jit_cache_pr

bringlein committed:
ruff....

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
fixing merge error

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>
